google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerOptions.java

```diff
@@ -134,6 +143,66 @@ public ServiceRpc create(SpannerOptions options) {
     }
   }
 
+  private static final AtomicInteger DEFAULT_POOL_COUNT = new AtomicInteger();
+
+  /** {@link ExecutorProvider} that is used for {@link AsyncResultSet}. */
+  interface CloseableExecutorProvider extends ExecutorProvider, AutoCloseable {
+    /** Overridden to suppress the throws declaration of the super interface. */
+    @Override
+    public void close();
+  }
+
+  static class FixedCloseableExecutorProvider implements CloseableExecutorProvider {
+    private final ScheduledExecutorService executor;
+
+    private FixedCloseableExecutorProvider(ScheduledExecutorService executor) {
+      this.executor = Preconditions.checkNotNull(executor);
+    }
+
+    @Override
+    public void close() {
+      executor.shutdown();
+    }
+
+    @Override
+    public ScheduledExecutorService getExecutor() {
+      return executor;
+    }
+
+    @Override
+    public boolean shouldAutoClose() {
+      return false;
+    }
+
+    /** Creates a FixedCloseableExecutorProvider. */
+    static FixedCloseableExecutorProvider create(ScheduledExecutorService executor) {
+      return new FixedCloseableExecutorProvider(executor);
+    }
+  }
+
+  /**
+   * Default {@link ExecutorProvider} for high-level async calls that need an executor. The default
+   * uses a cached thread pool containing a max of 8 threads. The pool is lazily initialized and
+   * will not create any threads if the user application does not use any async methods. It will
+   * also scale down the thread usage if the async load allows for that.
+   */
+  @VisibleForTesting
+  static CloseableExecutorProvider createDefaultAsyncExecutorProvider() {
+    return createAsyncExecutorProvider(8, 60L, TimeUnit.SECONDS);
+  }
+
+  @VisibleForTesting
+  static CloseableExecutorProvider createAsyncExecutorProvider(
+      int poolSize, long keepAliveTime, TimeUnit unit) {
+    String format = String.format("async-pool-%d-thread-%%d", DEFAULT_POOL_COUNT.incrementAndGet());
```

Comment:
Should we specify "spanner" in the thread prefix? It can be hard to troubleshoot where wild threadpools come from when an application uses multiple client libraries..

Reply:
Good point, done.